driver-reactive-streams/src/main/com/mongodb/reactivestreams/client/internal/BatchCursorFlux.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.reactivestreams.client.internal;
+
+import com.mongodb.lang.Nullable;
+import org.reactivestreams.Publisher;
+import org.reactivestreams.Subscriber;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.FluxSink;
+import reactor.core.publisher.Mono;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+class BatchCursorFlux<T> implements Publisher<T> {
+
+    private final BatchCursorPublisher<T> batchCursorPublisher;
+    private final Integer originalBatchSize;
+    private final Runnable resetBatchSize;
+    private final AtomicBoolean inProgress = new AtomicBoolean(false);
+    private final AtomicLong demandDelta = new AtomicLong(0);
+    private BatchCursor<T> batchCursor;
+    private FluxSink<T> sink;
+
+    BatchCursorFlux(final BatchCursorPublisher<T> batchCursorPublisher,
+                    @Nullable final Integer originalBatchSize,
+                    final Runnable resetBatchSize) {
+        this.batchCursorPublisher = batchCursorPublisher;
+        this.originalBatchSize = originalBatchSize;
+        this.resetBatchSize = resetBatchSize;
+    }
+
+    @Override
+    public void subscribe(final Subscriber<? super T> subscriber) {
+        Flux.<T>create(sink -> {
+            this.sink = sink;
+            sink.onRequest(l -> {
+                if (demandDelta.addAndGet(l) > 0 && inProgress.compareAndSet(false, true)) {
```

Comment:
I think there is a concurrency issue here.  Consider the case where an application calls Subscription.request multiple times.  The first call we will get demand and set inProgress to true.  Now imagine that the second call occurs just after line 57.  We've set the batchCursor but haven't recursed yet.  So the second call will observe demand and inProgress of false and will enter the conditional, then set inProgress to false and call recurseCursor.  Meanwhile the first call will also, concurrently, call recurseCursor.  

Also, batchCursor is not protected by a mutex and so I think it's not safe to assume that writes to it will be observed.

Reply:
Multiple calls to recurseCursor is fine as it contains a locking mechanism around inProgress - so if there is a race calling it via multiple calls to `request(n)` only one thread can action it.

> batchCursor is not protected by a mutex and so I think it's not safe to assume that writes to it will be observed.

Not sure I follow here, batchCursor is also protected by the inprogress check..

Reply:
@stIncMale can you check my reasoning here?

Reply:
### Scenario expressed by Jeff

> Consider the case where an application calls Subscription.request multiple times. The first call we will get demand and set inProgress to true. Now imagine that the second call occurs just after line 57. We've set the batchCursor but haven't recursed yet. So the second call will observe demand and inProgress of false and will enter the conditional, then set inProgress to false and call recurseCursor. Meanwhile the first call will also, concurrently, call recurseCursor.

@jyemin Yes, the method `recurseCursor` may be called concurrently in the scenario you describe. However, this method protects most of its body from being executed concurrently by using the same approach:
```java
if (inProgress.compareAndSet(false, true)) {//start executing mutually exclusive code in thread X
    ...
    doOnSuccess(results -> {
       ...
       inProgress.set(false);//end executing mutually exclusive code in thread Y
                             //(may not be the same as X, depending on how Reactor/BatchCursor work)
    })
}
```

A similar code in the `subscribe` method is meant to guarantee that:
1. `batchCursor` is initialized not more than once
2. `w_in_subscribe(batchCursor, bc)` happens-before (HB) any `r_in_recurseCursor(batchCursor)`.

#### HB(w_in_subscribe(batchCursor, <initial value>), r_in_closeCursor(batchCursor))?
`inProgress` does not provide HB between `w_in_subscribe(batchCursor, bc)` and `r_in_closeCursor(batchCursor)`. This HB may only be provided by the Reactor. @rozza Could you please confirm this HB relation is guaranteed?

Reply:
`closeCursor` can be called at any time via `sink::onDispose` & `sink::onCancel` - so the batch cursor may not be set.

Looking at the code there is a chance of the sink being cancelled before the batchCursor is set which currently wouldn't close the cursor:

```
batchCursorPublisher.batchCursor(batchSize).subscribe(bc -> {
                            batchCursor = bc;  // sink.isCancelled could be true at this point in time.
                            inProgress.set(false);
                            recurseCursor();
                        }, sink::error);
```

So I've added a check in `recurseCursor` - as `closeCursor` is idempotent it doesn't matter if it gets called multiple times.

Reply:
> `closeCursor` can be called at any time via `sink::onDispose` & `sink::onCancel`

And `closeCursor` may be called concurrently with the code we pass to the `onRequest` method, right? If that is the case, then making sure executions do not have data race due to writing to and reading from `batchCursor` is on us. The simplest way of achieving this is by marking `batchCursor` as `volatile`.

Reply:
Done.